kernel/mmanutils.c

```diff
@@ -194,106 +194,127 @@ static ssize_t _map_file_onto_memory(
     return ret;
 }
 
-/* ATTN-A: fix return types for this function */
-void* myst_mmap(
+long myst_mmap(
     void* addr,
     size_t length,
     int prot,
     int flags,
     int fd,
     off_t offset)
 {
-    void* ptr = (void*)-1;
-    int r;
+    long ret = -1;
+
+    /* fail if length is zero. Note that the page-alignment will
+     * be enforced by myst_mman_mprotect and myst_mman_mmap */
+    if (!length)
+        ERAISE(-EINVAL);
 
-    (void)flags;
+    // ATTN: Currently the check of addr is done by myst_mman_mprotect
+    // and myst_mman_mmap. Consider move the check here.
 
     /* check for invalid PROT bits */
     if (prot & (~MYST_PROT_MMAP_MASK))
-        return (void*)-1;
+        ERAISE(-EINVAL);
 
-    // Linux ignores fd when the MAP_ANONYMOUS flag is present
+    /* Linux ignores fd when the MAP_ANONYMOUS flag is present */
     if (flags & MAP_ANONYMOUS)
         fd = -1;
+    /* fail if fd is negative when the MAP_ANONYMOUS flag is not present */
+    else if (fd < 0)
+        ERAISE(-EBADF);
 
-    /* check file permissions upfront */
+    /* check file permissions and offset upfront */
     if (fd >= 0)
     {
         long flags;
         struct stat buf;
 
+        // ATTN: Use EBADF and EACCES for fd validation failures. This may not
+        // conform the Linux kernel behavior.
+
         /* fail if not a regular file */
         if (myst_syscall_fstat(fd, &buf) != 0 || !S_ISREG(buf.st_mode))
-            return (void*)-1;
+            ERAISE(-EBADF);
 
         /* get the file open flags */
         if ((flags = myst_syscall_fcntl(fd, F_GETFL, 0)) < 0)
-            return (void*)-1;
+            ERAISE(-EBADF);
 
         /* if file is not open for read */
         if ((flags & O_WRONLY))
-            return (void*)-1;
+            ERAISE(-EACCES);
 
         /* MAP_SHARED & PROT_WRITE set, but fd is not open for read-write */
         if ((flags & MAP_SHARED) && (prot & PROT_WRITE) && !(flags & O_RDWR))
-            return (void*)-1;
+            ERAISE(-EACCES);
+
+        /* fail is offset is not page-aligned */
+        if (offset % PAGE_SIZE)
+            ERAISE(-EINVAL);
     }
 
     if (fd >= 0 && addr)
     {
         ssize_t n;
+
         // ATTN: call mmap or mremap here so that this range refers to
         // a mapped region.
-        // ATTN: is non-page-aligned length valid?
+        // ATTN: Use EACCES for failures related to fd access. This may not
```

Comment:
LTP mmap15.c test tries to mmap to an address typically outside of app's linear address space -  `SYS_mmap(addr=fffffffffffff000 length=4096(1000) prot=1 flags=17 fd=3 offset=0)`, and expects EINVAL or ENOMEM error. The implementation here probably should check the whether the request address range is out of bound, before attempting to read file data (with mprotect()  call to make sure the memory range is writable).

Reply:
`myst_mman_mprotect` have done such checks and return EINVAL when the checks failed. I updated the PR to cover the case when `myst_mman_mprotect` is not called.